### PR TITLE
8339120: Use more fine-granular gcc unused warnings

### PR DESCRIPTION
**Pull request by magicus**
Currently, we issue -Wno-unused for all files in gcc, which is a rather big sledgehammer to get rid of some warnings that proliferate in a few areas of the build.

We should instead leave -Wunused turned on (as done by -Wall) and use a much more fine-grained approach to disabling specific warnings in specific files or libraries.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339120](https://bugs.openjdk.org/browse/JDK-8339120): Use more fine-granular gcc unused warnings (**Bug** - P3)

### Reviewers
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20733/head:pull/20733` \
`$ git checkout pull/20733`

Update a local copy of the PR: \
`$ git checkout pull/20733` \
`$ git pull https://git.openjdk.org/jdk.git pull/20733/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20733`

View PR using the GUI difftool: \
`$ git pr show -t 20733`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20733.diff">https://git.openjdk.org/jdk/pull/20733.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20733#issuecomment-2313687282)

### CODE REVIEW

**Comment by magicus** [comment]
If at some point the diff seems larger than needed, it could be that I have reordered things where the alphabetic order had capsized.

**Review by TheShermanTanker — APPROVED** [review]
That's a lot of new warnings, but I like the idea of being more fine grained in what is disabled and enabled

File: make/autoconf/flags-cflags.m4 [hunk 1]
```diff
@@ -235,14 +235,14 @@ AC_DEFUN([FLAGS_SETUP_WARNINGS],
       CFLAGS_WARNINGS_ARE_ERRORS="-Werror"
 
       # Additional warnings that are not activated by -Wall and -Wextra
-      WARNINGS_ENABLE_ADDITIONAL="-Wpointer-arith -Wsign-compare \
-          -Wunused-function -Wundef -Wunused-value -Wreturn-type \
-          -Wtrampolines"
+      WARNINGS_ENABLE_ADDITIONAL="-Wpointer-arith -Wreturn-type -Wsign-compare \
+          -Wtrampolines -Wundef -Wunused-const-variable -Wunused-function \
```
Comment (kimbarrett) [inline make/autoconf/flags-cflags.m4:239]: I think we don't want `-Wunused-const-variable=2` (eqv. `-Wunused-const-variable`) for C++
code. Recall that C++ const variables default to internal linkage (e.g. implicitly static). It's normal
to have a non-local constant in a header file that isn't used by every translation unit.  For C++ use
`-Wunused-const-variable=1`.  I think doing this might eliminate the need for disabling this warning
in a bunch of other places in this PR.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
We should make a similar set of changes for clang, though doing that in a separate
proposal is good.  Is there a JBS issue for that yet?

**Comment by magicus** [comment]
> We should make a similar set of changes for clang, though doing that in a separate
proposal is good. Is there a JBS issue for that yet?

Yes, we should. I am 80% done with that patch, but I have not yet opened a JBS ticket. Will do that now.

File: make/autoconf/flags-cflags.m4 (hunk 1 above)
Comment (magicus) [reply make/autoconf/flags-cflags.m4:239]: Good point, I'll try that.

Comment (magicus) [reply make/autoconf/flags-cflags.m4]: It turned out to be sort-of okay-ish. I explicitly listed like 6 or so per-file exclusions in Hotspot (even though my normal cutoff for just setting a component-wide exclude is 3-4) since it seems that you will want to look at and fix those.

File: make/modules/java.desktop/lib/ClientLibraries.gmk [hunk 2]
```diff
@@ -281,51 +281,51 @@ endif
 ################################################################################
 
 ifeq ($(USE_EXTERNAL_HARFBUZZ), true)
-   LIBFONTMANAGER_EXTRA_SRC =
-   LIBFONTMANAGER_LIBS += $(HARFBUZZ_LIBS)
-   LIBFONTMANAGER_CFLAGS += $(HARFBUZZ_CFLAGS)
+  LIBFONTMANAGER_EXTRA_SRC =
```
Comment (kimbarrett) [inline make/modules/java.desktop/lib/ClientLibraries.gmk:284]: I think this 3space -> 2space indentation change shouldn't be part of this PR, esp. since 3space is
used in other parts of this file.

Comment (magicus) [reply make/modules/java.desktop/lib/ClientLibraries.gmk:284]: Really? That is sooo broken. :( But fair enough, I'll do a separate PR on the broken indentation.

File: make/autoconf/flags-cflags.m4 (hunk 1 above)
Comment (magicus) [reply make/autoconf/flags-cflags.m4]: Actually, it was probably a bad idea. I've been chasing platform-specific build failures for the last hour or so. :( But now at least it is done. I hope you guys can use this to actually fix the problems so it was worth it. ;-)

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by TheShermanTanker — APPROVED** [review]

**Review by erikj79 — APPROVED** [review]
